### ISSUE TITLE
setActionHandler() should support GET and PUT actions

### ISSUE THREAD

**Issue by jenlincode** [eclipse/thingweb.node-wot#173]
Hello all!

For our setActionHandler() use cases, though we can use POST, technically it's syntax-wise incorrect for some of the methods we intend to use the setActionHandler on (some of them are GET requests to get sensor values or PUT requests to update values). 

May I get support for this to be updated?

**Comment by egekorkan** [comment]
Some remarks:

- setActionHandler is after the protocol handling so the code that would need to support this has to happen in the produce() stage.

- When you say "some of them are GET requests to get sensor values or PUT requests to update values", why don't you use a property to read the sensor values (HTTP GET) and have an action called updateSensorReading that will (probably) start communication with the GPS satellites. This would be an HTTP POST request since you are not writing data into a resource.

- If by "update" you mean that the request will supply the new values, you can write into the property by HTTP PUT.

**Comment by jenlincode** [comment]
Let me know if my understanding of the property usage is incorrect. I was hesitant to use it since the value would only update when the sensor calls a setHandlerAction() to update it -- but since the data associated with this value is large, I only wanted it to update if the frontend/client is requesting for it.  

So essentially, I didn't want the sensor to just broadcast updates unless it was requested. (It's not a value I would mark as observable.)

When you mention that the code that needs to support it needs to happen in the WoT.produce() stage, is there a setting to set "action" to GET somewhere I'm overlooking?

Also, when you say that you can write it into the property by HTTP PUT, I'm not intending on using it with a property. In this situation, no values are intended on being cached as properties, but just passed along as responses.

**Comment by egekorkan** [comment]
> When you mention that the code that needs to support it needs to happen in the WoT.produce() stage, is there a setting to set "action" to GET somewhere I'm overlooking?

Sorry that this was not clear, it was more meant for the person who would implement the update. If there would be way to do this as the user, it needs to happen at produce() stage. You haven't overlooked anything.

> Let me know if my understanding of the property usage is incorrect. I was hesitant to use it since the value would only update when the sensor calls a setHandlerAction() to update it -- but since the data associated with this value is large, I only wanted it to update if the frontend/client is requesting for it.

Now I understand the issue better. You can do something like the following:
https://github.com/eclipse/thingweb.node-wot/blob/0385fd9b456388348b9107028124de19f6b83231/examples/templates/exposed-thing/src/base.ts#L118
This way you can assign a handler to a property read operation that will update the value only when the HTTP GET request comes to the Thing.

With this issue, I also realize that the link to the Scripting API is not clear. Actually, you can find all the functions you can call at https://w3c.github.io/wot-scripting-api/. node-wot is an implementation of the Scripting API. @danielpeintner could you add information to this in the README?

**Comment by jenlincode** [comment]
Thank you for explaining that Ege! Also, can you share with me a more detailed example of how to use setPropertyReadHandler()? Admittedly, I find myself wishing there were more code snippets in the WoT API descriptions, as sometimes I find I learn better when the descriptions read a bit less clear for me.

Also, in retrospect, after thinking about it longer, I guess there's really nothing wrong with using an HTTP POST on a method that I meant for a get method. It just threw me a bit, syntax-wise.

**Comment by egekorkan** [comment]
You are right, there are explanations (https://www.w3.org/TR/wot-scripting-api/#the-setpropertyreadhandler-method) but no examples most of the time. 

The way to use it would be like in the previous comment where you can assign a function that will return a simple return value
```
thing.setPropertyReadHandler("myProperty", {
  // return sensor.value // here the sensor.value returns the value of the sensor
  // return 123;
}) 
```
or a promise

```
thing.setPropertyReadHandler("myProperty", {
	return new Promise((resolve, reject) => {
	    // read something
	    resolve(sensor.value); //this might be needed when the library is returning the value asynchronously
	});
}) 
```

**Comment by jenlincode** [comment]
Thank you, Ege!

**Comment by egekorkan** [comment]
@danielpeintner I would prefer to not close the issue since it is not possible to override the default protocol methods for the exposed things. This would however imply a substantial code structure change where the use case is not that wide spread.

**Comment by danielpeintner** [comment]
@egekorkan  What you are saying is that we should allow overriding the default protocol methods for exposed things?

**Comment by egekorkan** [comment]
I meant changing the invokeaction from POST to PUT if the user needs it. It could be relatively easy as a build option but maybe not so easy to specify it for different exposed things

**Comment by egekorkan** [comment]
At least it should be documented, maybe at the [Readme of the HTTP binding](https://github.com/eclipse/thingweb.node-wot/tree/master/packages/binding-http#server-example)?

**Comment by danielpeintner** [comment]
A easy way I can think of is to work on some *extensions* in form of a `nw` (node-wot) namespace. The node-wot runtime could be configured to listen to such settings.

Let's assume adding an action like this to a thing when producing it...

```
		actions: {
			increment: {
				description: "Incrementing counter value",
				"nw:opMethod": "PUT"
			}
```

The "flag" `"nw:opMethod": "PUT"` indicates to node-wot (`nw` namespace) that the action should be accepting PUT instead of POST for this specific action. (Note: this approach/annotation might affect *all* bindings/forms. Not sure if we want that.)

The problem/fear I have is that such a script works for node-wot but not for other runtimes. 
Does it make sense to specify it somewhere else?

Yet another solution could be to allow form entries when producing a thing and look at the provided settings. At the moment node-wot removes any form entry and creates them *new*.

**Comment by egekorkan** [comment]
I would say that node-wot specific information should not be passed from the scripting level but somehow in a config or before the WoT object is created. Otherwise, as @danielpeintner said, the same script would exhibit different behaviors in different implementations.

I think creating the custom TD would not be that difficult and can be done with some little changes to the code. However, handling the requests would change quite a bit since there are hard-coded parts in the handler, like [this](https://github.com/eclipse/thingweb.node-wot/blob/master/packages/binding-http/src/http-server.ts#L543). If for each request, the handler would look into the corresponding form and compare with the arrived request. Or with the wot-servient configuration, one can set the method for all the Things that will be created.
